Content.Shared/Body/Systems/SharedBloodstreamSystem.cs

```diff
@@ -364,31 +402,51 @@ public bool FlushChemicals(Entity<BloodstreamComponent?> ent, ProtoId<ReagentPro
     public bool TryModifyBloodLevel(Entity<BloodstreamComponent?> ent, FixedPoint2 amount)
     {
         if (!Resolve(ent, ref ent.Comp, logMissing: false)
-            || !SolutionContainer.ResolveSolution(ent.Owner, ent.Comp.BloodSolutionName, ref ent.Comp.BloodSolution))
+            || !SolutionContainer.ResolveSolution(ent.Owner, ent.Comp.BloodSolutionName, ref ent.Comp.BloodSolution, out var bloodSolution))
             return false;
 
+        var data = GetEntityBloodData(ent);
+
         if (amount >= 0)
-            return SolutionContainer.TryAddReagent(ent.Comp.BloodSolution.Value, ent.Comp.BloodReagent, amount, null, GetEntityBloodData(ent));
+            return SolutionContainer.TryAddReagent(ent.Comp.BloodSolution.Value, ent.Comp.BloodReagent, amount, null, data);
+
+        var bloodRef = new ReagentId(ent.Comp.BloodReagent, data);
 
-        // Removal is more involved,
-        // since we also wanna handle moving it to the temporary solution
-        // and then spilling it if necessary.
-        var newSol = SolutionContainer.SplitSolution(ent.Comp.BloodSolution.Value, -amount);
+        foreach (var (reagentId, _) in bloodSolution.Contents)
+        {
+            // Only remove our blood specifically.
+            if (reagentId != bloodRef)
+                continue;
+
+            SolutionContainer.RemoveReagent(ent.Comp.BloodSolution.Value, reagentId, -amount);
+            return true;
+        }
```

Comment:
this loop is extraneous and can be simplified to
```suggestion
    return SolutionContainer.RemoveReagent(ent.Comp.BloodSolution.Value, bloodRef, -amount) > FixedPoint2.Zero;
```